### PR TITLE
support datastore transaction options

### PR DESCRIPTION
**Pull request by neozwu**

### CODE REVIEW

**Comment by garrettjonesgoogle** [comment]
Removing myself as reviewer since this is WIP and @vam-google is secondary on Datastore.

**Comment by neozwu** [comment]
Kindly ping

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java [hunk 1]
```diff
@@ -118,4 +127,9 @@ public void rollback() {
   public Datastore getDatastore() {
     return datastore;
   }
+
+  @Override
+  public ByteString getTransactionId() {
+    return transaction;
```
Comment (vam-google) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java]: The name was chosen before this code change, but please rename this to `transactionId`.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 2]
```diff
@@ -62,22 +63,84 @@ public Batch newBatch() {
...
+          setPrevTransactionId(transaction);
+        }
+      }
+    }
+
+    void setPrevTransactionId(Transaction transaction) {
+      options =
+          options
+              .toBuilder()
+              .setReadWrite(
```
Comment (vam-google) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Please format this  somewhat prettier (one option to consider is to create local variables, for example for the value passed to `.setReadWrite()`).

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 3]
```diff
@@ -62,22 +63,84 @@ public Batch newBatch() {
...
+          transaction.rollback();
+        }
+        if (options.getModeCase().equals(TransactionOptions.ModeCase.READ_WRITE)) {
+          setPrevTransactionId(transaction);
+        }
+      }
+    }
+
+    void setPrevTransactionId(Transaction transaction) {
+      options =
```
Comment (vam-google) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: This seems not thread-safe.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/Datastore.java [hunk 4]
```diff
@@ -45,6 +53,17 @@
...
   }
 
+  /**
+   * A wrapper around {@link TransactionCallable} with {@link TransactionOptions}.
+   * If {@link TransactionOptions} mode is set to ReadWrite, its previous transaction id will
+   * be populated each time the underlying {@link TransactionCallable} is retried.
+   *
+   * @param <T> the type of the return value
+   */
+  interface ReadWriteTransactionCallable<T> {
```
Comment (vam-google) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/Datastore.java]: It does not look like this interface should be exposed. For example, both versions of runIntransaction() method (with and without options) accept `TransactionCallable<>`. It like the wrapper is an implementation detail (and should be). If that is the case, please remove it from the client's surface.

Comment (neozwu) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/Datastore.java]: I have removed `ReadWriteTransactionCallable` in the interface.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 5]
```diff
@@ -62,22 +63,84 @@ public Batch newBatch() {
...
+        transaction.commit();
+        return value;
+      } catch (Exception ex) {
+        transaction.rollback();
+        throw DatastoreException.propagateUserException(ex);
+      } finally {
+        if (transaction.isActive()) {
+          transaction.rollback();
+        }
+        if (options.getModeCase().equals(TransactionOptions.ModeCase.READ_WRITE)) {
```
[the remark this answers is not in the document]
Comment (neozwu) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Fixed

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 6]
```diff
@@ -62,22 +63,84 @@ public Batch newBatch() {
...
+          setPrevTransactionId(transaction);
+        }
+      }
+    }
+
+    void setPrevTransactionId(Transaction transaction) {
```
[the remark this answers is not in the document]
Comment (neozwu) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Done.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 3 above)
Comment (neozwu) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Fixed.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 2 above)
Comment (neozwu) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Done.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 7]
```diff
@@ -86,6 +149,20 @@ public T call() throws DatastoreException {
     }
   }
 
+  @Override
+  public <T> T runInTransaction(final TransactionCallable<T> callable, TransactionOptions options) {
```
[the remark this answers is not in the document]
Comment (neozwu) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Done.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java (hunk 1 above)
Comment (neozwu) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/TransactionImpl.java]: Done.

**Comment by neozwu** [comment]
PTAL @vam-google

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreHelperTest.java [hunk 8]
```diff
@@ -159,72 +153,4 @@ public void testFetchWithTransaction() throws Exception {
     assertEquals(entity2, values.get(1));
     verify(transaction);
   }
-
```
Comment (vam-google) [inline google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreHelperTest.java]: Why was this removed?

File: google-cloud-datastore/pom.xml [hunk 9]
```diff
@@ -41,7 +41,7 @@
     <dependency>
       <groupId>com.google.cloud.datastore</groupId>
       <artifactId>datastore-v1-proto-client</artifactId>
-      <version>1.4.1</version>
+      <version>1.5.1</version>
```
Comment (vam-google) [inline google-cloud-datastore/pom.xml]: should this match datastore-v1-protos (i.e. either both 1.5.0 or both 1.5.1)? I know it did not match before, but it seems like it was wrong originally.

Comment (garrettjonesgoogle) [reply google-cloud-datastore/pom.xml]: This version should just be removed at this point, it is in google-cloud-bom now.

File: google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreHelperTest.java (hunk 8 above)
Comment (neozwu) [reply google-cloud-datastore/src/test/java/com/google/cloud/datastore/DatastoreHelperTest.java]: After I cleaned up the surface, I got rid of the DatastoreHelper "default method", hence deleting the tests. The new impl logics are tested in DatastoreTest and ITs.

**Comment by garrettjonesgoogle** [comment]
@neozwu would you be able to finish this up?

File: google-cloud-datastore/pom.xml (hunk 9 above)
Comment (neozwu) [reply google-cloud-datastore/pom.xml]: Removed.

**Comment by neozwu** [comment]
@garrettjonesgoogle should be ready to merge once API goes public

**Comment by garrettjonesgoogle** [comment]
I added the "status: blocked" tag since this is blocked on the server side.
